docs/man/man1/msu.1.md
% msu(1)
%
% November 2025


NAME
----
msu - my shell utilities


SYNOPSIS
--------
*msu* require|install|uninstall 'MODULE'...

*msu* execute|install-many|uninstall-many 'PATH'

*msu* run 'MODULE'.'FUNC'

*msu* list

*msu* upgrade ['VERSION']

*msu* nuke

*msu* version ['MODULE']

*msu* where 'MODULE'

*msu* help ['MODULE']


DESCRIPTION
-----------

The msu(1) command adds useful utilities for use in your shell.
*msu* stands for 'My' 'Shell' 'Utilities'.


COMMANDS
--------

*h, help* ['MODULE']

   > Shows documentation of aliases of the specified module 'MODULE'.
   > If no module is specified, shows basic help information, shortened
   > for easier recap.

*i, install* 'MODULE'...

   > Installs the module 'MODULE' into the directory at
   > `${MSU_EXTERNAL_LIB}`. The module can be located either locally
   > or remotely.

   > If remote, a prefix shorthand is required to determine where to
   > fetch the module from. Available case-insensitive shorthands
   > include:

      bt: - from Bitbucket, using HTTPS e.g. 'bt:myusername/myrepo'
      bts: - from Bitbucket, using SSH e.g. 'bts:myusername/myrepo'
      gh: - from Github, using HTTPS e.g. 'gh:myusername/myrepo'
      ghs: - from Github, using SSH e.g. 'ghs:myusername/myrepo'
      gl: - for GitLab, using HTTPS e.g. 'gl:myusername/myrepo'
      gls: - for GitLab, using SSH e.g. 'gls:myusername/myrepo'

   > If the remote repository is private, SSH should be used
   > instead. This allows avoiding being prompted for password and
   > the request being rejected in cases where 2FA is enabled.

   > You can specify a version, using a git tag. For example,
   > 'gh:myusername/myrepo#v0.0.0' would fetch the remote module
   > at version 'v0.0.0', provided the git tag ('v0.0.0'; exactly
   > as specified) exists.

*im, install-many* 'PATH'

   > Installs modules listed in a file at path 'PATH'. Each line
   > of the file is assumed to be a single module and is eventually
   > passed to the *install* sub-command. This is useful when
   > bootstrapping machines.

*ls, list*

   > Lists the external modules you have installed.

*nk, nuke*

   > Uninstalls msu entirely from the machine. This involves
   > removing the msu executable, internal and external modules,
   > load strings added to ~/.bashrc and manpages.

*r, run* 'MODULE'.'FUNC'

   > Runs the function 'FUNC' in the module 'MODULE'. Any extra
   > arguments are passed to the function as is.

   > For example,

        $ msu run console.success 'a success message'

*re, require* 'MODULE'...

   > Loads the module 'MODULE' into a new bash environment,
   > making the modules' functions and variables available.
   > Usually, this environment should 'sourced' into the current
   > environment.

   > For example,

        $ . msu require console
        $ success 'a success message'

*u, uninstall* 'MODULE'...

   > Uninstalls the module 'MODULE' by removing its contents from
   > `${MSU_EXTERNAL_LIB}`.

*um, uninstall-many* 'PATH'

   > Uninstalls modules listed in a file at path 'PATH'. Each line
   > of the file is assumed to be a single module and is eventually
   > passed to the *uninstall* sub-command. This is useful in
   > un-bootstrapping machines (as a reverse of *install-many*).

*up, upgrade* ['VERSION']

   > Upgrades msu to the version 'VERSION', if passed. Otherwise
   > upgrades to the latest version. This request may hit the
   > network to check for new versions and execute downloads if
   > necessary.

*v, version* ['MODULE']

   > Shows version information for the module 'MODULE', if passed.
   > Otherwise, shows version information for msu itself.

*w, where* 'MODULE'

   > Prints the installation path of the external module 'MODULE'.
   > Exits with an error if module not installed.

*x, execute* 'PATH'

   > Executes the file at path 'PATH' by simply running the code
   > through bash, with the core library loaded before-hand. This
   > emulates the action of running a function with its body being
   > the contents of the file. This command should be considered
   > DANGEROUS as it is able to execute commands in a file without
   > the execution bit set.


ENVIRONMENT VARIABLES
---------------------

*`${MSU_EXTERNAL_LIB}`*

   > Set to the path to the directory holding the external modules.
   > It is where modules are installed to, with the *install*
   > and *install-many* sub-commands. It defaults to
   > `${HOME}/.msu`.


ALIASES
-------

On startup, aliases are loaded. Aliases added by msu are prefixed
with 'msu'. Remember any alias (including those added by msu itself)
can be overridden. See msu(3) for more details on how aliases work.

One of the most important aliases is *msu.reload*.

*msu.reload*

   > Reloads aliases from internal and external modules. This
   > allows you to load new aliases after installing modules.

Other aliases are listed in man(3).


SHEBANG
-------

msu can be used in a shebang line, which in almost all cases would
be '#!/usr/bin/env msu'. The shell script (using the shebang),
would receive the path to itself as `${1}` and have the other
arguments follow as normal, in `${2}`, `${3}`, ... This means
that the script can accept command-line arguments.


CAVEATS
-------

The commands *install*, *install-many* and *upgrade* use git for
cloning repositories and generating module metadata. Therefore, the
command git(1) must be available.



RESOURCES
---------

Source code: https://github.com/GochoMugo/msu

Issue tracker: https://github.com/GochoMugo/msu/issues


AUTHOR
------

*msu* is developed and maintained by Gocho Mugo.


COPYING
-------

THE MIT LICENSE (MIT)

Copyright \(C) 2015 Gocho Mugo \<mugo@forfuture.co.ke>


CHANGELOG.md

# Change Log

All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).


## Unreleased


## [0.7.0][0.7.0] - 2026-03-15

Added:

* `help` sub-command can show documentation of aliases
  in an external module.
* `install-many`, `uninstall` and `uninstall-many`
  sub-commands support the `--force` flag.
* Added bash completion.

Changed:

* `uninstall` fails if module does not exist, unless
  `--force` flag is used.


## [0.6.0][0.6.0] - 2026-03-09

Added:

* Added `where` sub-command.

Fixed:

* `install` command does not fail if installing from local path.
* `install` outputs error properly.
* `run` command supports interactive functions.
* `upgrade` command returns an exit code of zero if already at latest version.


## [0.5.0][0.5.0] - 2026-01-07

Added:

* Automatically checks for updates when some msu commands,
  such as `msu help`, are run.


## [0.4.0][0.4.0] - 2025-12-11

Revamped after years of neglect!


## [0.3.0][0.3.0] - 2016-03-19

Added:

* check dependencies aggressively, on installation

Removed:

* the `fs` and `net` modules are removed, to keep `msu` minimal


## [0.2.0][0.2.0] - 2016-02-26

Added:

* add command, `nuke`, for nuking `msu` (issue #19)
* add support for SSH in installing modules from private, remote repos (issue #18)
* allow assuming "yes" in a `yes_no` question, using `${MSU_ASSUME_YES}`

Fixed:

* fix using backspace as first key-press in password prompt
* fix unnecessary logging to a file named `log` when using `msu` in a shebang
* fix listing external modules, when directory at `${MSU_EXTERNAL_LIB}` does **not** exist
* fix listing internal modules i.e. only list the `*.sh` files


## [0.1.0][0.1.0] - 2016-02-13

Added:

* add the alias `msu.reload` for reloading aliases
* add support for use in shebang e.g `#!/usr/bin/env msu`

Changed:

* prefix all aliases added by msu library with `msu`

Fixed:

* Fix path to executable, if executed directly (not through symlink)


## [0.0.0][0.0.0] - 2016-01-24

This is the very first version of `msu`.


<!-- Release links are placed here for easier updating -->
[0.0.0]:https://github.com/GochoMugo/msu/releases/tag/0.0.0
[0.1.0]:https://github.com/GochoMugo/msu/releases/tag/0.1.0
[0.2.0]:https://github.com/GochoMugo/msu/releases/tag/0.2.0
[0.3.0]:https://github.com/GochoMugo/msu/releases/tag/0.3.0
[0.4.0]:https://github.com/GochoMugo/msu/releases/tag/0.4.0
[0.5.0]:https://github.com/GochoMugo/msu/releases/tag/0.5.0
[0.6.0]:https://github.com/GochoMugo/msu/releases/tag/0.6.0
[0.7.0]:https://github.com/GochoMugo/msu/releases/tag/0.7.0


lib/msu.sh
#!/usr/bin/env bash
# msu (my-shell-utils)
#
# Copyright (c) 2015 GochoMugo <mugo@forfuture.co.ke>


#set -e errexit


MSU_EXE="${BASH_SOURCE[0]}"
MSU_REAL_EXE=$(readlink -f "${MSU_EXE}")
MSU_LIB=$(dirname "${MSU_REAL_EXE}") # directory holding our library
export MSU_EXE
export MSU_LIB


# modules
# shellcheck source=lib/core.sh
source "${MSU_LIB}/core.sh" # so LOW-LEVEL
msu_require "metadata"      # how you like me now?


# warn user, if we are running as root/superuser.
msu_run core_utils.is_superuser && {
  echo
  echo " !!         you are running as SUPERUSER        !! "
  echo " !! with much power, comes great responsibility !! "
  echo " !!           you have been WARNED!             !! "
  echo
}


# parse command line arguments
case "${1:-''}" in
  "env" )
    msu__load
  ;;
  "re" | "require" )
    msu_require "${2:-''}"
  ;;
  "r" | "run" )
    msu_run "${@:2}"
  ;;
  "x"  | "execute" )
    msu_execute "${2:-''}" "${@:3}"
  ;;
  "i" | "install" )
    msu_run core_utils.install "${@:2}"
    msu_run core_utils.update_check
  ;;
  "im" | "install-many" )
    msu_run core_utils.install_from_list "${2}"
    msu_run core_utils.update_check
  ;;
  "u" | "uninstall" )
    msu_run core_utils.uninstall "${@:2}"
    msu_run core_utils.update_check
  ;;
  "um" | "uninstall-many" )
    msu_run core_utils.uninstall_from_list "${2}"
    msu_run core_utils.update_check
  ;;
  "up" | "upgrade" )
    msu_run core_utils.upgrade "${2}"
  ;;
  "ls" | "list" )
    msu_run core_utils.list_modules "${2}"
    msu_run core_utils.update_check
  ;;
  "nk" | "nuke" )
    msu_run core_utils.nuke
  ;;
  "h" | "help" )
    if [ "${2}" ]
    then
      msu_run core_utils.show_help "${2}"
      msu_run core_utils.update_check
    else
      echo
      echo " msu by ${MSU_AUTHOR_NAME} <${MSU_AUTHOR_EMAIL}>"
      echo
      echo " usage:"
      echo "    msu require|install|uninstall <mod>..."
      echo "    msu execute|install-many|uninstall-many <path>"
      echo "    msu run <mod>.<func>"
      echo "    msu list"
      echo "    msu upgrade [version]"
      echo "    msu nuke"
      echo "    msu version [mod]"
      echo "    msu where <mod>"
      echo "    msu help [mod]"
      echo
      echo " commands:"
      echo "    h  | help            show help information of module [mod] or this help info"
      echo "    i  | install         install the module(s) <mod>..."
      echo "    im | install-many    install from module-list at path <path>"
      echo "    ls | list            list installed modules"
      echo "    nk | nuke            nuke msu entirely"
      echo "    r  | run             run the function <func> in module <mod>"
      echo "    re | require         require the library module <mod>"
      echo "    u  | uninstall       uninstall the module(s) <mod>..."
      echo "    um | uninstall-many  uninstall from module-list at path <path>"
      echo "    up | upgrade         upgrade to the latest version"
      echo "    v  | version         show version information of module [mod] or msu itself"
      echo "    w  | where           show path to an external module"
      echo "    x  | execute         execute file at the path <path>"
      echo
      echo " for more help information, see msu(1)"
      echo
      echo " see https://github.com/GochoMugo/msu/issues for bug reporting"
      echo " and feature requests"
      echo
      msu_run core_utils.update_check
    fi
  ;;
  "v" | "version" )
    if [ "${2}" ]
    then
      msu_run core_utils.show_metadata "${2}"
    else
      echo "  version   ${MSU_VERSION}"
      echo "  build     ${MSU_BUILD_HASH:-?}"
      echo "  date      ${MSU_BUILD_DATE:-?}"
    fi
    msu_run core_utils.update_check
  ;;
  "w" | "where" )
    msu_run core_utils.where "${2}"
    msu_run core_utils.update_check
  ;;
  * )
    # maybe, we are being used in a shebang e.g. #!/usr/bin/env msu
    if [ "${1}" ]
    then
        FILE="$(readlink -f "${1}" 2> /dev/null)"
        if [ -r "${FILE}" ]
        then
            msu_execute "${FILE}" "${@:2}"
        fi
    fi
    # otherwise, we just ignore it!
  ;;
esac
